server/src/main/java/org/opensearch/common/document/DocumentField.java

```diff
@@ -75,6 +90,110 @@ public DocumentField(String name, List<Object> values) {
         this.values = Objects.requireNonNull(values, "values must not be null");
     }
 
+    public DocumentField(byte[] in) throws IOException {
```

Comment:
The protobuf should not leak out I believe, this is `@PublicApi` by and large, applies to other classes as well (like SearchHit ...). I don't have a single option to suggest here, but it is clear that current implementation is very much coupled with serialization (StreamInput & Co), the possible paths I see here are:
 - mirror `DocumentField` for protobuf  (that would also allow to properly implement `BytesWriteable.Writeable` since it does not)
 - completely detach the serde logic form `DocumentField` into supporting classes (fe `DocumentFieldSerDe` or `DocumentFieldFactory`, ....)

@andrross @saratvemulapalli do you have thoughts here?

Reply:
Sorry, it took me awhile to truly understand what is happening here, but the rename to BytesWriteable makes it clear: the fact that this code makes the assumption that any byte array is serialized protobuf breaks the abstraction that was described with the BytesWriteable discussion. How would this code work if another serialization mechanism was introduced?

I think the clean way to implement this would be to detach serde from these OpenSearch model classes. That would mean regardless of the serde mechanism in use, the same (protocol-agnostic) model class would be constructed, and those model classes wouldn't leak any implementation details. That would mean in the case of protobuf you would deserialize the bytes from the wire into a protobuf class, and then you would transform that protobuf class into the POJO with no protobuf dependencies. There will performance implications however, as you'll be potentially creating more garbage on the Java heap to clean up with the additional transformation.

Ultimately I agree with @reta that it's not really a tenable design to embed parallel protobuf classes as fields in these model classes. From a design perspective it makes for very bimodal behavior of the class and doesn't follow object-oriented patterns. From a more practical perspective it leaks protobuf dependencies far and wide outside of the server implementation.

Reply:
Thank you @reta and @andrross! I attempted detaching the serde for `DocumentField` as an example in this commit: https://github.com/VachaShah/OpenSearch/commit/9ec5c260129b8cc474636ddae9e735fdb455d09b. Let me know what you think.